ui/src/adapters/parsers/forms.ts

```diff
@@ -145,11 +160,21 @@ export const credentialFormParser = getStrictParser<
       issueCredential: issueCredentialFormDataParser,
     })
     .transform(({ issuanceMethod, issueCredential }, context) => {
-      const { credentialExpiration, credentialSubject, proofTypes } = issueCredential;
+      const { credentialExpiration, credentialSubject, proofTypes, refreshService } =
+        issueCredential;
       const { type } = issuanceMethod;
 
+      if (refreshService.enabled && credentialExpiration === undefined) {
+        context.addIssue({
+          code: z.ZodIssueCode.custom,
+          fatal: true,
+          message: `Credential expiration must set when refresh service is enabled.`,
+        });
+      }
```

Comment:
The fact that you need to add an error in the transformation phase just by checking the values (not transforming them in a way that can fail) is an indicator that the parser is incorrect.

`transform` is a map function that is called only when the parsing succeeds and when the parsing succeeds no further checking on the values should be required.

Reply:
Agree. Validation moved to form level.